fix(android): Fix auth flow and callback thread safety, and pass `fd` through FFI

jamilbk opened the pull request:
* Refactor sharedPreferences to only save the AccountId
* Update TeamId -> AccountId to match naming elsewhere
* Update JWT -> Token to avoid confusion; this token is **not** a valid JWT and should be treated as an opaque token
* Update FFI `connect` to accept an optional file descriptor (int32) as a first argument. This seemed to be the most straightforward way to pass it to the tunnel stack. Retrieving it via callback is another option, but retrieving return vars with the `jni` was more complex. We could have used a similar approach that we did in the Apple client (enumerating all fd's in the `new()` function until we found ours) but this approach is [explicitly documented/recommended](https://developer.android.com/reference/android/net/VpnService.Builder#establish()) by the Android docs so I figured it's not likely to break.

Additionally, there was a thread safety bug in the recent JNI callback implementation that consistently crashed the VM with `JNI DETECTED ERROR IN APPLICATION: use of invalid jobject...`. The fix was to use `GlobalRef` which has the explicit purpose of outliving the `JNIEnv` lifetime so that no `static` lifetimes need to be used.

jamilbk committed:
Update portal url and jwt variable names and usage

* Rename JWT -> Token
* Rename team-id -> account-id
* Refactor sharedPreferences to save only the accountId and build auth
  URL on the fly

jamilbk committed:
Pass optional file descriptor through FFI connect

jamilbk commented on kotlin/android/app/build.gradle:
```diff
@@ -32,10 +32,10 @@ android {
             // Debug Config
             manifestPlaceholders.hostName = "app.firez.one"
-            buildConfigField("String", "AUTH_HOST", "\"localhost\"")
+            buildConfigField("String", "AUTH_HOST", "\"10.0.2.2\"")
```
This is the host's IP as seen by the android debugging environment.

jamilbk replied:
Nevermind, using `app.firez.one` since testing locally with Docker is problematic

jamilbk committed:
Checkpoint refactor Onboarding -> Settings

jamilbk committed:
Fix weird string bug

jamilbk committed:
Fix incorrect splash layout

jamilbk committed:
Allow connections for websocket

jamilbk committed:
Don't use different callback handlers for prod vs dev

jamilbk committed:
Fix auth flow

jamilbk committed:
Fix auth URLs to staging

jamilbk commented on kotlin/android/app/build.gradle:
```diff
@@ -32,10 +32,10 @@ android {
+            buildConfigField("String", "AUTH_HOST", "\"app.firez.one\"")
+            buildConfigField("String", "AUTH_SCHEME", "\"https\"")
+            buildConfigField("Integer", "AUTH_PORT", "443")
+            buildConfigField("String", "CONTROL_PLANE_URL", "\"wss://app.firez.one/\"")
```
@pratikvelani You should now have an account on staging you can use to sign in with the magic link method, which has real Gateways, and a live connlib working with it. Use `pratik@spark.dev` to sign in.

With this change you no longer need to run `PortalMock`.

jamilbk commented on kotlin/android/app/src/main/AndroidManifest.xml:
```diff
@@ -8,6 +8,12 @@
 
+    <queries>
+        <intent>
+            <action android:name="android.support.customtabs.action.CustomTabsService" />
```
```suggestion
            <action android:name="android.support.customtabs.action.CustomTabsService" />
```

This fixed a crash I was hitting with CustomTabs.

jamilbk commented on kotlin/android/app/src/main/AndroidManifest.xml:
```diff
@@ -51,7 +57,7 @@
                 <category android:name="android.intent.category.BROWSABLE" />
 
-                <data android:scheme="https" android:host="${hostName}" />
+                <data android:scheme="https" android:host="${hostName}" android:path="/handle_client_auth_callback" />
```
```suggestion
                <data android:scheme="https" android:host="${hostName}" android:path="/handle_client_auth_callback" />
```

Without the `path` specifier here, the AppLink would intercept the initial `sign_in` path, preventing the user from actually signing in.

jamilbk commented on kotlin/android/app/src/main/java/dev/firezone/android/features/applink/ui/AppLinkHandlerActivity.kt:
```diff
@@ -27,11 +27,13 @@ class AppLinkHandlerActivity : AppCompatActivity(R.layout.activity_app_link_hand
             when (action) {
                 is AppLinkViewModel.ViewAction.AuthFlowComplete -> {
-                    // Continue with onboarding
+                    // TODO: Continue starting the session showing sessionFragment
```
@pratikvelani This is where I left off. We need to show the session fragment instead of the `AUTH Complete` TextView here.

jamilbk commented on kotlin/android/app/src/main/java/dev/firezone/android/features/session/ui/SessionViewModel.kt:
```diff
@@ -18,6 +18,7 @@ internal class SessionViewModel @Inject constructor(
 
     fun startSession() {
         viewModelScope.launch {
+            
```
```suggestion
```

jamilbk committed:
Update kotlin/android/app/src/main/java/dev/firezone/android/features/session/ui/SessionViewModel.kt

Signed-off-by: Jamil <jamilbk@users.noreply.github.com>

jamilbk commented on kotlin/android/app/src/main/java/dev/firezone/android/features/session/backend/SessionManager.kt:
```diff
@@ -18,13 +19,14 @@ internal class SessionManager @Inject constructor(
 
+            if (config.accountId != null && config.token != null) {
                 sessionPtr = Session.connect(
+                    -1, // TODO: get fd from VpnService. See VpnBuilder#establish().detachFd()
```
@conectado Here is where it would be convenient to pass in the file descriptor in a single line

jamilbk commented on kotlin/android/app/src/main/java/dev/firezone/android/util/CustomTabsHelper.kt:
@pratikvelani This was taken from the CustomTabs demo repo. I think we should support other browsers that the user may have set as default (e.g. Firefox)

jamilbk committed:
Fix clippy error

jamilbk committed:
Remove extra whitespace

jamilbk commented on elixir/apps/web/lib/web/endpoint.ex:
```diff
@@ -39,6 +39,7 @@ defmodule Web.Endpoint do
 
   socket "/live", Phoenix.LiveView.Socket,
     websocket: [
+      check_origin: :conn,
```
@andrewdryga this fixed liveview issues in development when accessing the portal by my Mac's local LAN IP. Unsure if we want this on `prod` however?

AndrewDryga replied:
That's actually a good catch and we do want it on prod, thank you

jamilbk commented on elixir/config/runtime.exs:
```diff
@@ -81,12 +81,6 @@ if config_env() == :prod do
-  config :web, Web.Auth,
-    platform_redirect_urls: %{
-      "apple" => "firezone://handle_client_auth_callback",
-      "android" => "#{external_url_scheme}://#{external_url_host}/handle_client_auth_callback"
```
@AndrewDryga These actually don't change between environments. The App Link is only intercepted for `app.firez.one`

jamilbk committed:
Add parameter stub to windows tunnel

jamilbk committed:
Add note Re: Android tunnel DRYing

pratikvelani commented on kotlin/android/app/src/main/java/dev/firezone/android/features/settings/ui/SettingsFragment.kt:
```diff
@@ -7,27 +7,28 @@ import android.view.inputmethod.EditorInfo
 
 @AndroidEntryPoint
-internal class OnboardingFragment : Fragment(R.layout.fragment_onboarding) {
+internal class SettingsFragment : Fragment(R.layout.fragment_settings) {
```
@jamilbk I was implementing a separate SettingsFragment, as it would have some UI for the user to delete the token and in debug mode a UI for us to view app logs.

jamilbk replied:
Ah, but wouldn't `Sign out` accomplish deleting the token? Wiping the app's data would too right? Confused why that would be a dedicated action in the Settings fragment.

A debug log viewer in the SettingsFragment sounds good. Do you want to merge this and add your changes into it, or the other way around?

The main thing here was renaming `OnboardingFragment` since that's not really what this view is.

pratikvelani replied:
I want to move the `Sign out` button to the settings screen too. We can merge this PR, I will push a separate PR for it.

jamilbk replied:
Sounds good. This is one is already large enough.

jamilbk committed:
feat(android): session and services setup (#1936)

Opening a draft to diff @pratikvelani's WIP branch

---------

Co-authored-by: Pratik Velani <pratikvelani@gmail.com>

jamilbk committed:
Bypass CSRF

pratikvelani committed:
Added debug user sign in and vpn service builder

jamilbk committed:
Bypass auth to test tunnel

jamilbk committed:
Use detachFd() since we'll be closing the fd ourselves

jamilbk committed:
Fix JNI callbacks thread safety. tl;dr: use GlobalRef

A JObject with static lifetime was incorrectly used to maintain a
reference to the callback class instance in the Java VM. This pointer
was being GC'd, causing an "invalid jobject frame reference" exception
to occur.

The fix, as clearly laid out in the [JNI crate docs
here](https://docs.rs/jni/latest/jni/objects/struct.GlobalRef.html),
is to obtain a global reference to the callback JObject which is
guaranteed by the VM not to be GC'd until explicitly freed.

jamilbk committed:
Thanks clippy

jamilbk committed:
Remove Android Studio's mess

jamilbk committed:
Use webpki certs for android client

jamilbk committed:
Use non-repo token; Fix typo in ctrl plane URL

jamilbk committed:
Debug control plane connection... 422

pratikvelani committed:
Added default token value for release build

jamilbk committed:
Use attach_current_thread_as_daemon

jamilbk committed:
Minor tunnel fixes; Update WebRTC to hopefully fix Android issue

jamilbk committed:
Add note re: MTU

jamilbk committed:
revert webpki-roots to native-roots for phoenix-channel

conectado commented on rust/connlib/clients/android/src/lib.rs:
```diff
@@ -301,23 +307,34 @@ fn connect(
     portal_url: JString,
     portal_token: JString,
-    callback_handler: JObject<'static>,
+    callback_handler: JObject,
```
Just a note to myself in the future:

The problem here was that an owned `JObject` can never have an `'static` lifetime since by default `JObject`s have a lifetime of the frame that's calling the ffi function and the only way to obtain an static one is through a `GlobalRef` which only implements `AsRef<JObject<'static>>`.

conectado commented on rust/connlib/clients/android/src/lib.rs:
```diff
@@ -301,23 +307,34 @@ fn connect(
+    let callback_handler = env
+        .new_global_ref(callback_handler)
+        .map_err(CallbackError::NewGlobalRefFailed)
+        .unwrap();
```
```suggestion
let Ok(callback_handler) = env.new_global_ref(callback_handler) else { return std::ptr::null() };
```

jamilbk replied:
Fixed

conectado commented on rust/connlib/clients/android/src/lib.rs:
```diff
@@ -301,23 +307,34 @@ fn connect(
 #[no_mangle]
-pub unsafe extern "system" fn Java_dev_firezone_connlib_Session_connect(
-    mut env: JNIEnv<'static>,
+pub extern "system" fn Java_dev_firezone_android_tunnel_TunnelSession_connect(
```
this is still `unsafe` since the pointers here might be unvalid.

conectado replied:
Also now we need to assume that `fd` is a vaild file descriptor.

jamilbk replied:
Fixed

conectado commented on rust/connlib/clients/android/src/lib.rs:
```diff
@@ -76,7 +81,7 @@ impl CallbackHandler {
 fn call_method(
     env: &mut JNIEnv,
-    this: &JObject,
+    this: &GlobalRef,
```
This can still be a `&JObject`, it doesn't implement neither `Send` nor `Sync` so we can be sure that it's not moved to another thread.

jamilbk replied:
Fixed

conectado commented on rust/connlib/clients/android/src/lib.rs:
```diff
@@ -46,7 +49,9 @@ impl Clone for CallbackHandler {
 pub enum CallbackError {
-    #[error("Failed to attach current thread as daemon: {0}")]
+    #[error("Failed to obtain a global reference: {0}")]
+    NewGlobalRefFailed(#[source] jni::errors::Error),
```
We do nothing with this error so we can remove it.

jamilbk replied:
Removed

conectado commented on rust/connlib/libs/tunnel/src/lib.rs:
```diff
@@ -504,6 +506,9 @@ where
                     match dev.device_channel.mtu().await {
+                        // XXX: Do we need to fetch the mtu every time? In most clients it'll
+                        // be hardcoded to 1280, and if not, it'll only change before packets start
+                        // to flow.
```
But it could still happen that it changes, if that happens this could panic.

jamilbk replied:
I see. Pardon my noob question, but won't adding syscalls to the hotpath take up a lot of CPU time / overhead?

I ask because WireGuard-go [stores the MTU on the device struct](https://github.com/WireGuard/wireguard-go/blob/469159ecf7d108dc1d44f8789dedbb463b7bd3a6/device/device.go#L86), and boringtun [does something similar](https://github.com/cloudflare/boringtun/blob/e1d6360d6ab4529fc942a078e4c54df107abe2ba/boringtun/src/device/mod.rs#L382).

If so, I think it's an acceptable edge case for this to possibly crash if an admin changes the tunnel interface MTU while it's up.

jamilbk replied:
From there, if it is crashing and we do find this to be an unacceptable UX in the wild, we could refresh it periodically, or maybe it's possible to subscribe to the MTU set syscall?

conectado replied:
Now that I thought about this, I think that the `read` call will not fail if the buffer is small it will just read a mangled packet(the rest of this loop expects a full packet).

So we can just do as boringtun does that it reads the mtu once and updates it periodically. Still calling a method here is fine, the `DeviceChannel` can store the mtu.

jamilbk replied:
Gotcha, that makes sense. Thanks for clarifying! Glad it wasn't a stupid question.

conectado commented on rust/connlib/libs/gateway/src/control.rs:
```diff
@@ -144,13 +144,15 @@ impl<CB: Callbacks + 'static> ControlPlane<CB> {
 impl<CB: Callbacks + 'static> ControlSession<IngressMessages, CB> for ControlPlane<CB> {
     #[tracing::instrument(level = "trace", skip(private_key, callbacks))]
     async fn start(
+        fd: Option<i32>,
```
We are going to change it later so it doesn't matter but I think it's better if we expect here a `RawFd`.

jamilbk replied:
Yeah, this thought was already crossing my mind as I wrote this, so I'll note it down for when we change it!

conectado commented on rust/connlib/libs/tunnel/src/tun_android.rs:
```diff
@@ -28,30 +29,33 @@ impl Drop for IfaceDevice {
+        // We stick with a hardcoded MTU of 1280 for now. This could be improved by
+        // finding the MTU of the underlying physical interface and subtracting 80
+        // from it for the WireGuard overhead, but that's a lot of complexity
+        // for little gain.
```
we still should find the actual mtu or at least have the callbacks tell us the mtu that it used at some point, since if it configured a great MTU we could read packets partially.

conectado replied:
also note that this should be the wireguard's interface mtu not the physical interface, each peer could use a different physical interface and it can also change.

jamilbk replied:
It's a little involved in Android since we first have to retrieve the interface name with `TUNGETIFF`. But I'll take that on as an exercise for the reader.

jamilbk replied:
I'll save this for the next PR if that's alright

conectado commented on rust/connlib/libs/tunnel/src/tun_android.rs:
```diff
@@ -28,30 +29,33 @@ impl Drop for IfaceDevice {
-        } else {
-            Ok(Self { fd })
-        }
+    pub async fn new(fd: Option<i32>) -> Result<Self> {
```
This is actually unsafe since `fd` might not be valid and we assume it's vaild in the `read`/`write` implementations. Since we are going to change how we obtain the `fd` it's not that important right now but we need to remember that for the refactor.

jamilbk replied:
Added

jamilbk replied:
Aaaand removed so this will compile

conectado commented on rust/connlib/libs/common/Cargo.toml:
```diff
@@ -31,8 +31,14 @@ rand = { version = "0.8", default-features = false, features = ["std"] }
 parking_lot = "0.12"
 
+# Needed for Android logging until tracing is working
+log = "0.4"
```
I think we should enable the [`log` feature](https://docs.rs/tracing/latest/tracing/#crate-feature-flags) from the tracing crate instead. then we can just drop `log` entirely and `tracing` will automatically emit `log`'s logs if we are on android.

jamilbk replied:
Nice, it looks like we can also consume the log events from our dependencies and consume those in tracing as well with https://docs.rs/tracing/latest/tracing/#consuming-log-records

Perhaps I'll save that change for another PR?

conectado reviewed (dismissed):
Looks good! left some comments that'd be good if we can fix before merging.

jamilbk committed:
Add PR feedback minimally

jamilbk committed:
Fix merge conflict semantic error

pratikvelani commented on rust/connlib/libs/tunnel/src/device_channel_unix.rs:
```diff
@@ -60,8 +60,8 @@ impl DeviceChannel {
-pub(crate) async fn create_iface() -> Result<(IfaceConfig, DeviceChannel)> {
-    let dev = Arc::new(IfaceDevice::new().await?.set_non_blocking()?);
+pub(crate) async fn create_iface(fd: Option<i32>) -> Result<(IfaceConfig, DeviceChannel)> {
+    let dev = Arc::new(IfaceDevice::new(fd).await?.set_non_blocking()?);
```
```suggestion
    let dev = Arc::new(unsafe { IfaceDevice::new(fd).await?.set_non_blocking()? });
```

Fixes CI build issue
```rust
error[E0133]: call to unsafe function is unsafe and requires unsafe function or block
  --> connlib/libs/tunnel/src/device_channel_unix.rs:64:24
   |
64 |     let dev = Arc::new(IfaceDevice::new(fd).await?.set_non_blocking()?);
   |                        ^^^^^^^^^^^^^^^^^^^^ call to unsafe function
   |
   = note: consult the function's documentation for information on how to avoid undefined behavior
```

jamilbk replied:
Fixed by removing`unsafe` from the function definition.

jamilbk committed:
revert unsafe

conectado commented on rust/phoenix-channel/Cargo.toml:
```diff
@@ -16,3 +16,4 @@ url = "2.4.0"
 serde_json = "1.0.104"
 thiserror = "1.0.47"
 tokio = { version = "1.28.2", features = ["net", "time"] }
+log = "0.4"
```
```suggestion
```

`phoenix-channel` is only used by the relay, it doesn't need `log`.

jamilbk committed:
Update rust/phoenix-channel/Cargo.toml

Co-authored-by: Gabi <gabrielalejandro7@gmail.com>
Signed-off-by: Jamil <jamilbk@users.noreply.github.com>

conectado reviewed (approved):
Left a few extra non-blocking comments.

jamilbk committed:
Remove log from phoenix-channel
